### PR TITLE
C++: use a native implementation of timeout

### PR DESCRIPTION
**Pull request by cmnrd**
This pulls in https://github.com/lf-lang/reactor-cpp/pull/36 which extends reactor-cpp with a native timeout handler. This PR adjusts the code generator accordingly and removes the old Timeout mechanism. The timeout policy implemented by this PR is to invoke shutdown precisely at tag  (timeout, 0).

Note that this PR does not provide a solution for https://github.com/lf-lang/lingua-franca/issues/1496 as further discussion on the semantics of timeout is required. Thus, PR is only the first step and provides a simple timeout implementation as this is needed to move forward with the enclave feature. We can refine the policy in future PRs.

Fixes https://github.com/lf-lang/lingua-franca/issues/725

### CODE REVIEW

File: org.lflang/src/lib/cpp/time_parser.hh [hunk 1]
```diff
@@ -87,30 +107,30 @@ std::stringstream &operator>>(std::stringstream& in, reactor::Duration& dur) {
...
-               unit == "minute" || unit == "minutes") {
+               unit == "minute" || unit == "minutes" || unit == "m") {
       std::chrono::duration<double, std::ratio<60, 1>> tmp{value};
       dur = std::chrono::duration_cast<reactor::Duration>(tmp);
-    } else if (unit == "hour" || unit == "hours") {
+    } else if (unit == "hour" || unit == "hours" || unit == "h") {
       std::chrono::duration<double, std::ratio<3600, 1>> tmp{value};
       dur = std::chrono::duration_cast<reactor::Duration>(tmp);
-    } else if (unit == "day" || unit == "days") {
+    } else if (unit == "day" || unit == "days" || unit == "?") {
```
Comment (oowekyala) [inline org.lflang/src/lib/cpp/time_parser.hh:130]:
```suggestion
    } else if (unit == "day" || unit == "days" || unit == "d") {
```

File: org.lflang/src/lib/cpp/time_parser.hh [hunk 2]
```diff
@@ -131,6 +151,8 @@ std::string validate_time_string(const std::string& time) {
     return "The empty string is not a valid time!";
   } else if (trimmed[0] == '-') {
     return "Negative values are not a valid time!";
+  } else if (iequals("forever", time)) {
+    return ""; // "forever" is a valid value
```
Comment (oowekyala) [inline org.lflang/src/lib/cpp/time_parser.hh:155]: Is the value "forever" defined by LF or by the C++ target only? It's the first time I see this.

File: org.lflang/src/org/lflang/generator/cpp/CppAssembleMethodGenerator.kt [hunk 3]
```diff
@@ -123,14 +120,17 @@ class CppAssembleMethodGenerator(private val reactor: Reactor) {
...
-    private fun assembleReaction(reaction: Reaction) = with(PrependOperator) {
-        """
+    private fun assembleReaction(reaction: Reaction): String {
+        val sources = reaction.sources.filter { it.variable is Port }
+        return with(PrependOperator) {
+            """
             |// ${reaction.name}
         ${" |"..reaction.triggers.joinToString(separator = "\n") { declareTrigger(reaction, it) }}
-        ${" |"..reaction.sources.joinToString(separator = "\n") { declareDependency(reaction, it) }}
+        ${" |"..sources.joinToString(separator = "\n") { declareDependency(reaction, it) }}
```
Comment (oowekyala) [inline org.lflang/src/org/lflang/generator/cpp/CppAssembleMethodGenerator.kt:129]: Can one not use an action as a source? Or is it handled somewhere else?

Comment (cmnrd) [reply org.lflang/src/org/lflang/generator/cpp/CppAssembleMethodGenerator.kt:129]: Yes, actions can be sources, but we don't need to tell this the runtime (the runtime only keeps track of the reaction that an action triggers). They are only brought into scope elsewhere.

File: org.lflang/src/lib/cpp/time_parser.hh (hunk 2 above)
Comment (cmnrd) [reply org.lflang/src/lib/cpp/time_parser.hh:155]: Forever is also used in the C target. I am not sure if it is specified anywhere to be honest.

File: org.lflang/src/org/lflang/generator/cpp/CppExtensions.kt [hunk 4]
```diff
@@ -1,13 +1,7 @@
...
 
 import org.eclipse.emf.ecore.resource.Resource
-import org.lflang.InferredType
-import org.lflang.TargetProperty
-import org.lflang.TimeUnit
-import org.lflang.indexInContainer
-import org.lflang.isBank
-import org.lflang.isGeneric
-import org.lflang.isMultiport
+import org.lflang.*
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/cpp/CppExtensions.kt:4]: I believe we don't allow wildcard imports in our style guide.

**Review by lhstrh — APPROVED** [review]
:+1:

File: org.lflang/src/org/lflang/generator/cpp/CppExtensions.kt (hunk 4 above)
Comment (cmnrd) [reply org.lflang/src/org/lflang/generator/cpp/CppExtensions.kt:4]: For Kotlin we refer to the [Coding Conventions](https://kotlinlang.org/docs/coding-conventions.html) which don't specify this. Wild card imports are quite idiomatic in Kotlin and are applied automatically by the IntelliJ auto-formatter.

**Pull request merged**